### DISCUSSION TITLE
Rules are not processed for channels if  they values are  modified by alexa (using option68 on)

### DISCUSSION THREAD

**Discussion by johnder1972** [arendst/Tasmota#11621] [Ideas]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._
Hello, 

When setting a rule like "on channel1 do var1 %value% endon" the rule is processed only if the changes for channel are made with the web interface of with the console. But If an Alexa device is linked to the channel  , when changing the value with alexa, the value of the web interface change (after refresh screen) but the rule is not processed. (I made my test with two channels attached.)
But by using a simple dimmer (without option68) it's works... but only for one dimmer :-(

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): chinaese clone of sonoff pro
- [ ] Tasmota binary firmware version number used: 9.3.1.2 dev
  - [ ] Pre-compiled
  - [x] Self-compiled April 6 2021 17h45
- [x] Flashing tools used: OTA
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:

17:03:22.991 CMD: Backlog Template; Module; GPIO 255
17:03:22.994 SRC: WebConsole from 192.168.1.19
17:03:22.998 CMD: Group 0, Index 1, Command "BACKLOG", Data "Template; Module; GPIO 255"
17:03:23.068 SRC: Backlog
17:03:23.071 CMD: Group 0, Index 1, Command "TEMPLATE", Data ""
17:03:23.076 RSL: RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
17:03:23.274 SRC: Backlog
17:03:23.278 CMD: Group 0, Index 1, Command "MODULE", Data ""
17:03:23.282 RSL: RESULT = {"Module":{"0":"Generic"}}
17:03:23.488 SRC: Backlog
17:03:23.492 CMD: Group 0, Index 1, Command "GPIO", Data "255"
17:03:23.498 RSL: RESULT = {"GPIO0":{"32":"Button1"},"GPIO1":{"0":"None"},"GPIO2":{"417":"PWM2"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO9":{"33":"Button2"},"GPIO10":{"34":"Button3"},"GPIO12":{"0":"None"},"GPIO13":{"416":"PWM1"},"GPIO14":{"35":"Button4"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}
17:03:26.078 WIF: Checking connection...
```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:

17:05:16.802 CMD: Group 0, Index 1, Command "BACKLOG", Data "Rule1; Rule2; Rule3"
17:05:16.942 SRC: Backlog
17:05:16.946 CMD: Group 0, Index 1, Command "RULE", Data ""
17:05:16.951 RSL: RESULT = {"Rule1":"OFF","Once":"OFF","StopOnError":"OFF","Length":43,"Free":468,"Rules":""}
17:05:17.154 SRC: Backlog
17:05:17.157 CMD: Group 0, Index 2, Command "RULE", Data ""
17:05:17.162 RSL: RESULT = {"Rule2":"ON","Once":"OFF","StopOnError":"OFF","Length":67,"Free":444,"Rules":"**on Channel1 do var1 %value% endon on Channel2 do var2 %value% endon**"}
17:05:17.366 SRC: Backlog
17:05:17.369 CMD: Group 0, Index 3, Command "RULE", Data ""
17:05:17.374 RSL: RESULT = {"Rule3":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}

```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
Status 0
17:06:31.483 SRC: WebConsole from 192.168.1.19
17:06:31.486 CMD: Group 0, Index 1, Command "STATUS", Data "0"
17:06:31.492 RSL: STATUS = {"Status":{"Module":0,"DeviceName":"Tasmota","FriendlyName":["Tasmota1","Tasmota2"],"Topic":"tasmota_CFBC02","ButtonTopic":"0","Power":3,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
17:06:31.523 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:09:51","StartupUTC":"2021-04-06T15:56:40","Sleep":50,"CfgHolder":4617,"BootCount":32,"BCResetTime":"2021-04-06T13:32:19","SaveCount":175,"SaveAddress":"F6000"}}
17:06:31.555 RSL: STATUS2 = {"StatusFWR":{"Version":"9.3.1.2(tasmota)","BuildDateTime":"2021-04-06T15:42:15","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8285","CR":"395/699"}}
17:06:31.573 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":4,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["Dorothy",""],"TelePeriod":300,"Resolution":"55C180C0","SetOption":["00008001","2805C8000100060000005A0A000000000000","00840080","00006000","00000000"]}}
17:06:31.601 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":596,"Free":404,"Heap":27,"ProgramFlashSize":1024,"FlashSize":1024,"FlashChipId":"144051","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6"}}
17:06:31.633 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota_CFBC02-7170","IPAddress":"192.168.1.58","Gateway":"192.168.1.1","Subnetmask":"255.255.255.0","DNSServer":"192.168.1.1","Mac":"C8:2B:96:CF:BC:02","Webserver":2,"WifiConfig":4,"WifiPower":17.0}}
17:06:31.655 RSL: STATUS7 = {"StatusTIM":{"UTC":"2021-04-06T16:06:31","Local":"2021-04-06T17:06:31","StartDST":"2021-03-28T02:00:00","EndDST":"2021-10-31T03:00:00","Timezone":"+01:00","Sunrise":"06:16","Sunset":"19:27"}}
17:06:31.673 RSL: STATUS10 = {"StatusSNS":{"Time":"2021-04-06T17:06:31"}}
17:06:31.681 RSL: STATUS11 = {"StatusSTS":{"Time":"2021-04-06T17:06:31","Uptime":"0T00:09:51","UptimeSec":591,"Heap":27,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":0,"POWER1":"ON","Channel1":21,"POWER2":"ON","Channel2":98,"Color":"36FA","Wifi":{"AP":1,"SSId":"Dorothy","BSSId":"A6:91:B1:86:DA:80","Channel":11,"RSSI":80,"Signal":-60,"LinkCount":1,"Downtime":"0T00:00:05"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:

when changing with web interface :


17:07:28.387 CMD: Group 0, Index 1, Command "CHANNEL", Data "12"
17:07:28.392 RSL: RESULT = {"POWER1":"ON","Channel1":12,"POWER2":"ON","Channel2":98,"Color":"1FFA"}
17:07:28.404 RUL: CHANNEL1 performs "var1 12"
17:07:28.407 SRC: Rule
17:07:28.410 CMD: Group 0, Index 1, Command "VAR", Data "12"
17:07:28.415 RSL: RESULT = {"Var1":"12"}
17:07:28.424 RUL: CHANNEL2 performs "var2 98"
17:07:28.427 SRC: Rule
17:07:28.430 CMD: Group 0, Index 2, Command "VAR", Data "98"
17:07:28.435 RSL: RESULT = {"Var2":"98"}
17:07:28.844 CFG: Saved to flash at F5, Count 176, Bytes 4096

When changing with alexa : 

17:08:26.008 WIF: Checking connection...
17:08:33.269 HTP: Hue API (/cfbc02/lights/217825313/state) from 192.168.1.6
17:08:33.273 HTP: Hue POST args ({"on":true,"bri":196})
17:08:33.280 SRC: Hue
17:08:33.285 RSL: RESULT = {"POWER1":"ON"}
17:08:33.289 RSL: POWER1 = ON
17:08:33.292 RSL: RESULT = ON
17:08:33.297 RUL: No valid JSON (ON)
17:08:33.299 RUL: No valid JSON (ON)
17:08:33.302 HTP: Hue Result ([{"success":{"/lights/217825313/state/on":true}},{"success":{"/lights/217825313/state/bri":196}}])
17:08:33.337 HTP: Hue API (/cfbc02/lights/217825313) from 192.168.1.6
17:08:33.341 /lights path=/lights/217825313
17:08:33.350 HTP: Hue Result ({"state":{"on":true,"bri":196,"alert":"none","effect":"none","reachable":true},"type":"Extended color light","n...
17:08:33.847 CFG: Saved to flash at F4, Count 177, Bytes 4096
17:08:40.400 HTP: Hue API (/cfbc02/lights/217825313) from 192.168.1.6
17:08:40.404 /lights path=/lights/217825313
17:08:40.414 HTP: Hue Result ({"state":{"on":true,"bri":196,"alert":"none","effect":"none","reachable":true},"type":"Extended color light","n...
17:08:40.443 HTP: Hue API (/cfbc02/lights/217825313) from 192.168.1.6
17:08:40.447 /lights path=/lights/217825313
17:08:40.456 HTP: Hue Result ({"state":{"on":true,"bri":196,"alert":"none","effect":"none","reachable":true},"type":"Extended color light","n...
17:08:40.488 HTP: Hue API (/cfbc02/lights/217825313) from 192.168.1.6



```

### TO REPRODUCE
_Steps to reproduce the behavior:_

Put 2 pwn 
setoption68 1
configure friendynames for alexa
create rules like :

rule2
on Channel1 do var1 %value% endon
on Channel2 do var2 %value% endon

test separates channels control bars with webui, rules are responding (ok)
test to change value with alexa, webui is changing but rules are not responding

(for additional context , ? may be the scope of this behavior is bigger ?)
Change end of rule's lines by "break" 
set channel1 by command console -> ok
set  channel2 by command console -> rule is doing something strange (the rule's trigger mix between channels)

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

rules should response as they response to the webui

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._
![image](https://user-images.githubusercontent.com/62481212/113744536-16517580-9705-11eb-9c67-d8e9424f9681.png)

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

Also, I can see taht by using "break" in place of "endon" for most than one line in the rule. The rule is not matching to good line when processing.
Just have a test a see the console. I should open another isuue for this but it's the same context.

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by s-hadinger** [comment]
The reason is that when the change comes from Alexa, no light status is published.

If I understand well, you would like a complete light status published after each change?

**Reply by johnder1972** [reply]
Exact, I think also that the problem is coming cause light status is not published. A line like "CMD: Group 0, Index 1, Command "CHANNEL", Data "12"
17:07:28.392 RSL: RESULT = {"POWER1":"ON","Channel1":12,"POWER2":"ON","Channel2":98,"Color":"1FFA"}"
 should be published also when alexa control.

?? ... or alexa shoud start the matching command (??as the webui does??)  ??

When Alexa drive a simple dimmer, rules are working. 

I don't know if rules are working when alexa drive a dimmer+ct+... light

**Comment by johnder1972** [comment]
Hello, 
To be more precise : 
The reason why I use rules is because I want to control my shutters with Alexa. For example, with a **dual relay module**, the way that I found for controlling % of tasmota shutters with Alexa is to create a third component as pwm( seen as a dimmer) And then I control this dimmer with alexa and I created rules to set shutters positions following  de dimmer's position. 
I also add a "one button" control by rules for button1 . You can find my rules here : (I pass some other config like unlink buttons. If you want more info I have a draft doc in French)

For this I hide the two first friendy names and use the 3th only with Alexa

```lua

Rule1
On button1#state do ShutterStopToggleDir1 break  
on Shutter1#Position do Backlog
 Rule2 0;
 Dimmer %value%;
 Rule2 1
Break



Rule2
on Dimmer#State do Backlog
 ShutterPosition %value% ;
Break


```
And it works like a charm. 

.But by trying doing it with a four relays module, it seems a bit more complex  :-)  ( channel in place of simple dimmer, rules more complex, ...)

But I think that the best way is to enhance Tasmota shutters. Indeed, when using setoption80,  if only one friendly name could be created for each shutter (so for two relays) and if the shutter is shown to Alexa like a dimmable light (not like two power lights) and if Alexa is able to control directly the shutters position following the received "bri" value, it should be better and more universal.

What do you think about ?

thx
john

**Reply by johnder1972** [reply]
sorry for the number of edits

**Comment by oponyx** [comment]
Hi,
please anybody found a solution for this?
